src/main/java/org/folio/event/handler/ItemCreateAsyncRecordHandler.java

```diff
@@ -0,0 +1,113 @@
+
+  private static final Logger log = LogManager.getLogger();
+
+  @Autowired
+  private PieceService pieceService;
+
+  public ItemCreateAsyncRecordHandler(Context context, Vertx vertx) {
+    super(vertx, context);
+    SpringContextUtil.autowireDependencies(this, context);
+  }
+
+  @Override
+  public Future<String> handle(KafkaConsumerRecord<String, String> kafkaConsumerRecord) {
+    if (log.isDebugEnabled()) {
+      log.debug("ItemCreateAsyncRecordHandler::handle, kafkaRecord={}", kafkaConsumerRecord.value());
+    }
+
+    var payload = new JsonObject(kafkaConsumerRecord.value());
+
+    String eventType = payload.getString("type");
+    if (!Objects.equals(eventType, DomainEventPayloadType.CREATE.name())) {
+      log.info("processItemCreateEvent:: unsupported event type: {}", eventType);
+      return Future.succeededFuture();
+    }
+
+    if (!validatePayload(payload)) {
+      log.warn("processItemCreateEvent:: payload validation failed");
+      return Future.succeededFuture();
+    }
+
+    try {
+      var tenantId = extractTenantFromHeaders(kafkaConsumerRecord.headers());
+      var dbClient = new DBClient(getVertx(), tenantId);
+      return processItemCreationEvent(payload, dbClient)
+        .onSuccess(v -> log.info("ItemCreateAsyncRecordHandler::handle, event processed successfully"))
+        .onFailure(t -> log.error("Failed to process event: {}", kafkaConsumerRecord.value(), t))
+        .map(kafkaConsumerRecord.key());
+    } catch (Exception e) {
+      log.error("Failed to process export history kafka record, kafkaRecord={}", kafkaConsumerRecord, e);
+      return Future.failedFuture(e);
+    }
+  }
+
+  private Future<Void> processItemCreationEvent(JsonObject payload, DBClient dbClient) {
+    var itemObject = payload.getJsonObject("new");
+    var itemId = itemObject.getString("id");
+
+    return pieceService.getPiecesByItemId(itemId, dbClient)
+      .compose(pieces -> updatePieces(pieces, itemObject, dbClient));
+  }
+
+  private Future<Void> updatePieces(List<Piece> pieces, JsonObject itemObject, DBClient client) {
+    if (CollectionUtils.isEmpty(pieces)) {
+      log.info("updatePieces:: no pieces to update found, nothing to update");
+      return Future.succeededFuture();
+    }
+
+    var holdingId = itemObject.getString("holdingsRecordId");
+    var tenantId = client.getTenantId();
+    var updateRequiredPieces = filterPiecesToUpdate(pieces, holdingId, tenantId);
+    updatePieceFields(updateRequiredPieces, holdingId, tenantId);
+    return pieceService.updatePieces(updateRequiredPieces, client);
+  }
+
+  private List<Piece> filterPiecesToUpdate(List<Piece> pieces, String holdingId, String tenantId) {
+    return pieces.stream()
+      .filter(piece ->
+        !StringUtils.equals(piece.getReceivingTenantId(), tenantId)
+          && (Objects.isNull(piece.getLocationId()) || !StringUtils.equals(piece.getLocationId(), holdingId)))
```

Comment:
(Objects.isNull(piece.getLocationId()) || !StringUtils.equals(piece.getLocationId(), holdingId))
should be 
(Objects.isNull(piece.getLocationId()) && !StringUtils.equals(piece.getHoldingId(), holdingId))

Reply:
by the way, this check with holdingId will not help in scope of this story, because holding ids would be the same after changing ownership to another tenant, because we have this story: https://folio-org.atlassian.net/browse/MODINV-1074 

But this check with holdingId will be useful for our story from next sprint when we chaning holding acrross single tenant. In this case holding ids would be different

Reply:
Thanks, update filter to skip same tenantId, holdingId or existingLocationId